### PR TITLE
optimize checks

### PR DESCRIPTION
**Pull request by robieta**
This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59393**
* #59392
* #59391
* #59389

### CODE REVIEW

File: tools/autograd/templates/VariableType.h [hunk 1]
```diff
@@ -42,14 +42,16 @@ using at::Quantizer;
 using ConstQuantizerPtr = const c10::intrusive_ptr<Quantizer>&;
 using c10::optional;
 
+// TODO: Once we have C++17, this can be a template function rather than a macro.
+#define UNPACK_TENSOR( t, pos ) \
+  TORCH_CHECK( t.defined(), "Expected a Tensor of type Variable but found an undefined Tensor for argument #", pos, " '", #t, "'" ); \
```
Comment (swolchok) [inline tools/autograd/templates/VariableType.h:47]: use compile-time string literal concatenation -- `" '" #t "'"`

Comment (swolchok) [reply tools/autograd/templates/VariableType.h:47]: if pos is a literal we should stringify it at compile-time too

File: torch/csrc/autograd/VariableTypeUtils.h [hunk 2]
```diff
@@ -76,13 +76,23 @@ inline void throw_error_out_requires_grad(const char* name) {
...
-inline void throw_error_for_complex_autograd(const Tensor& tensor, const char* name) {
-  if (tensor.requires_grad()) {
-    TORCH_CHECK(!tensor.is_complex(), name,
-                " does not support automatic differentiation for outputs with complex dtype.");
+inline bool is_complex_autograd(const Tensor& tensor) {
+  return tensor.is_complex() && tensor.requires_grad();
+}
+
+inline bool is_complex_autograd(const TensorList& tensorlist) {
+  for (auto tensor: tensorlist) {
```
Comment (swolchok) [inline torch/csrc/autograd/VariableTypeUtils.h:84]: shouldn't this be `const auto&`?

File: torch/csrc/autograd/VariableTypeUtils.h [hunk 3]
```diff
@@ -76,13 +76,23 @@ inline void throw_error_out_requires_grad(const char* name) {
...
+    if (is_complex_autograd(tensor)) {
+      return true;
+    }
   }
+  return false;
 }
 
+// TODO: Once we have C++17, this can be a template function rather than a macro.
+#define THROW_ERROR_FOR_COMPLEX_AUTOGRAD( t, name ) \
+  TORCH_CHECK( !is_complex_autograd(t), name, " does not support automatic differentiation for outputs with complex dtype." );
```
Comment (swolchok) [inline torch/csrc/autograd/VariableTypeUtils.h:94]: if name is supposed to be a literal then use compile-time string literal concatenation -- remove the comma

File: torch/csrc/autograd/FunctionsManual.h [hunk 4]
```diff
@@ -32,6 +32,9 @@ struct IndexRangeGenerator {
     size_t i = 0;
 };
 
+inline bool isFwGradDefined(const Tensor& t) {
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.h:35]: Why is this added? The overload below already covers that no?
Also it would be nice to reduce code in header files...

File: torch/csrc/autograd/VariableTypeManual.cpp [hunk 5]
```diff
@@ -38,38 +38,7 @@ C10_EXPORT std::vector<at::DeprecatedTypeProperties*> allCUDATypes() {
...
-
-Variable & checked_cast_variable(Tensor & t, const char * name, int pos) {
-  if (!t.defined()) {
-    AT_ERROR("Expected a Tensor of type Variable but found an undefined Tensor for argument #", pos, " '", name, "'");
-  }
-  return t;
-}
-}
-
-const Tensor & unpack(const Tensor & t, const char * name, int pos) {
```
Comment (albanD) [inline torch/csrc/autograd/VariableTypeManual.cpp:57]: Are these functions part of the c++ API? Is this BC-breaking?

File: torch/csrc/autograd/VariableTypeUtils.h [hunk 6]
```diff
@@ -76,13 +76,23 @@ inline void throw_error_out_requires_grad(const char* name) {
...
+  for (auto tensor: tensorlist) {
+    if (is_complex_autograd(tensor)) {
+      return true;
+    }
   }
+  return false;
 }
 
+// TODO: Once we have C++17, this can be a template function rather than a macro.
+#define THROW_ERROR_FOR_COMPLEX_AUTOGRAD( t, name ) \
```
Comment (albanD) [inline torch/csrc/autograd/VariableTypeUtils.h:93]: I'm not sure to follow what the benefit of having a macro is here? Since you do more function calls anyways?

File: tools/autograd/templates/VariableType.h [hunk 7]
```diff
@@ -42,14 +42,16 @@ using at::Quantizer;
 using ConstQuantizerPtr = const c10::intrusive_ptr<Quantizer>&;
 using c10::optional;
 
+// TODO: Once we have C++17, this can be a template function rather than a macro.
+#define UNPACK_TENSOR( t, pos ) \
```
Comment (albanD) [inline tools/autograd/templates/VariableType.h:46]: Do we really need a macro for this given that it is only used in the codegen anyways? You can replace the macro in the template with these two lines.

This used to be much more complex before the Tensor/Variable merge, but not much is left.

File: tools/autograd/gen_inplace_or_view_type.py [hunk 8]
```diff
@@ -170,13 +173,10 @@ def unpack_args(f: NativeFunction) -> Tuple[List[str], List[Binding]]:
             continue
 
         is_tensor_list = is_tensor_list_type(binding.argument.type)
-        ref = (not is_nullable) and not is_tensor_list
-        suffix = '_opt' if is_nullable and not is_tensor_list else ''
-        body.append(UNPACK_TENSOR.substitute(
+        template = (UNPACK_TENSOR_LIST if is_tensor_list else UNPACK_TENSOR)
```
Comment (albanD) [inline tools/autograd/gen_inplace_or_view_type.py:176]: I'm not sure how this is equivalent to the old code, in particular, I don't see the `_list` magic there.
Did you diff the generated code to make sure it is still the same?

File: torch/csrc/autograd/FunctionsManual.h (hunk 4 above)
Comment (robieta) [reply torch/csrc/autograd/FunctionsManual.h:35]: The issue is that with just the optional overload we wrap `t` (due to the implicit conversion), and then immediately check that it is defined and then unwrap it on the other side. That turns out to be a non-trivial amount of added work. (Same with inlining; it's a small function on the hot path so we want it to be inlined.)

File: torch/csrc/autograd/VariableTypeManual.cpp (hunk 5 above)
Comment (robieta) [reply torch/csrc/autograd/VariableTypeManual.cpp:57]: Not to my knowledge. (@ezyang?) If they are, we will still want to make them call the macros and directly use the macros in our own code for perf.

File: tools/autograd/templates/VariableType.h (hunk 7 above)
Comment (robieta) [reply tools/autograd/templates/VariableType.h:46]: I elected for the macro because I didn't want the manually defined functions to drift from the code-gen'd ones.

File: tools/autograd/gen_inplace_or_view_type.py (hunk 8 above)
Comment (robieta) [reply tools/autograd/gen_inplace_or_view_type.py:176]: I didn't. (I'll make a note to do so.) list magic is now handled by using a different codegen template. The key point for understanding equivalence is that there are no instances of unpack_opt, so the logic simplifies.

File: tools/autograd/templates/VariableType.h (hunk 7 above)
Comment (albanD) [reply tools/autograd/templates/VariableType.h:46]: They should be updated the same way. But it is know that whenever we change the codegen, the same update should be done on the manually defined functions.

File: tools/autograd/templates/VariableType.h (hunk 1 above)
Comment (robieta) [reply tools/autograd/templates/VariableType.h]: Done.

File: torch/csrc/autograd/VariableTypeUtils.h (hunk 3 above)
Comment (robieta) [reply torch/csrc/autograd/VariableTypeUtils.h]: Done.
